ibis/backends/clickhouse/compiler/values.py

```diff
@@ -1098,6 +1098,13 @@ def _fmt(op, _name: str = _name, **kw):
 del _fmt, _name, _op
 
 
+@translate_val.register(ops.ExtractMicrosecond)
+def _extract_microsecond(op, **kw):
+    arg = translate_val(op.arg, **kw)
+    dtype = serialize(op.output_dtype)
+    return f"CAST(substring(formatDateTime({arg}, '%f'), 1, 3) AS {dtype})"
```

Comment:
```suggestion
    return f"CAST(formatDateTime({arg}, '%f') AS {dtype})"
```

You don't need the `substring` call, the return value of `formatDateTime` here is exactly what needs to be cast:

```
weebill :) select now64(6), formatDateTime(now64(6), '%f');

┌───────────────────now64(6)─┬─formatDateTime(now64(6), '%f')─┐
│ 2023-06-22 06:28:12.497311 │ 497311                         │
└────────────────────────────┴────────────────────────────────┘
```

Reply:
This is a bit more complicated, pushing up a fix now.